pr-review: rate-limited reviews are never auto-retried (no equivalent of dev-lead-retry)

## Problem

When the pr-review agent runs while its advisory bots are rate-limited / out of quota (e.g. Codex "usage limits", CodeRabbit "review limit reached"), the advisory-gate correctly **withholds** donpetry-bot's auto-approval rather than approving prematurely (the behavior from #657/#658). The run completes "successfully" but posts no approval — leaving the PR at `REVIEW_REQUIRED` with green CI and `mergeable_state: blocked`.

Nothing then auto-recovers that PR once the rate limits reset. dev-lead has a dedicated remediator for exactly this class of bug (`dev-lead-retry.yml` — scans for `status=rate-limited` markers and re-dispatches when the limit clears), but **pr-review has no equivalent**.

### Observed instance

On PR #683 the pr-review run at `02:48Z` withheld approval because Codex was out of quota and CodeRabbit was rate-limited. ~10 hours and ~10 `pr-review-sweep` passes later, the PR still had **no** donpetry-bot approval. It only merged after a **manual** `pr-review-trigger.yml` dispatch (`force_review=true`) once the limits had reset. So the existing hourly sweep is demonstrably **not** recovering the rate-limited case.

### Why the existing sweep doesn't cover it

`pr-review-sweep.yml` re-dispatches PRs that are `REVIEW_REQUIRED` + **CI green** + **not already reviewed at head** (built for the event-gap/stuck case, #573). A rate-limited-and-withheld review falls outside that net, most likely because:

1. the withheld run left an idempotency marker, so the sweep treats the PR as "already reviewed at head" and skips it; and/or
2. the sweep's "CI green" gate sees sibling `cancelled`/`failed` check runs (e.g. concurrency-cancelled `dev-lead / dispatch`) and concludes CI isn't green, so it skips.

Root cause to confirm in `scripts/review-one-pr.sh` / `scripts/sweep-stuck-reviews.sh` / `scripts/engine.sh` (marker-write behavior on a rate-limited bail, and the sweep's green-CI predicate).

## Options

**Option 1 — Make `pr-review-sweep` rate-limit-aware (preferred).**
Treat a PR whose most recent pr-review run withheld approval due to `status=rate-limited` (advisory-gate) as retryable — i.e. ignore the "already reviewed at head" marker for that specific case, and don't let unrelated `cancelled`/`failed` sibling checks suppress the re-dispatch. Lowest cost, reuses the existing dispatch path, no new workflow.

**Option 2 — Add `pr-review-retry.yml` mirroring `dev-lead-retry.yml`.**
A dedicated scheduled scanner keying on `status=rate-limited` markers left by the pr-review cascade, re-dispatching `pr-review-trigger.yml` (as the bot PAT) once the limit clears. Most consistent with the dev-lead pattern and cleanly separates concerns, at the cost of a second sweep-like workflow.

### Recommendation

Lean **Option 1**: it closes the gap with the least surface area and keeps a single re-trigger path. Pivot to Option 2 only if the marker/idempotency semantics make the rate-limited case hard to express cleanly inside the existing sweep.

## Acceptance criteria

- A pr-review run that withholds approval specifically because advisory bots were rate-limited leaves a machine-detectable marker (`status=rate-limited`) on the PR.
- After the limit window resets, the PR is automatically re-reviewed (no manual `force_review` dispatch) and donpetry-bot can approve.
- Sibling `cancelled`/`failed` checks that are unrelated to the actual CI gate do not suppress the retry.
- Default behavior for non-rate-limited PRs is unchanged (no extra reviews, idempotency preserved).
- Covered by a bats test mocking the rate-limited marker + green CI and asserting a re-dispatch.

## Context / references

- `dev-lead-retry.yml`, `scripts/dev-lead-retry.sh` — the proven retry pattern.
- `pr-review-sweep.yml`, `scripts/sweep-stuck-reviews.sh` — existing stuck-review remediator (#573).
- `#657` / `#658` — advisory-gate: pr-review never auto-approves when advisory bots are rate-limited/out-of-quota.
- Discovered while monitoring the merge of #683.

_Filed by Claude Code while monitoring PR #683._

## Dev-Lead Implementation Report — issue #711

**Status:** complete (Option 1 — make `pr-review-sweep` rate-limit-aware). Changes are local only; no commit/push/PR per the engagement rules.

### What changed
The withhold-due-to-advisory-bot-rate-limit path (introduced by #657/#658) correctly keeps donpetry-bot from auto-approving while Codex/CodeRabbit are out of quota — but it left **no marker**, so once CI settled green no event re-fired and the PR stranded at `REVIEW_REQUIRED`. Two coordinated pieces close the gap:

1. **`scripts/lib/advisory-review-gate.sh`** — two new pure-ish helpers:
   - `detect_advisory_rate_limit <reviews/comments json>`: returns 0 when a known advisory/review bot's **latest** submission body matches a focused rate-limit phrase pattern (so a newer real review supersedes an older notice, and vice versa). Detection set is the gate's `ADVISORY_BOTS` **plus `coderabbitai`** (CodeRabbit posts an explicit out-of-quota comment but isn't gated on). Non-bot authors are ignored.
   - `maybe_post_rate_limited_marker <pr_url> <head_sha> <reset_iso> <comments_json>`: posts a **deduplicated** marker comment. The marker prefix (`rate-limited` before `v1`) deliberately never matches the idempotency regex `<!-- pr-review-agent v1 sha=…`.

2. **`scripts/review-one-pr.sh`** — in the non-`FORCE_REVIEW` withhold branch (gate rc=1), before `exit 100`, an isolated subshell sources the gate and, **if** the withhold is rate-limit-driven, stamps:
   ```
   <!-- pr-review-agent rate-limited v1 sha=<HEAD> status=rate-limited reset=<ISO> -->
   ```
   `reset` = now + `RATE_LIMIT_RETRY_BACKOFF` (default 3600s). Non-rate-limited withholds are untouched.

3. **`scripts/sweep-stuck-reviews.sh`** — a rate-limit retry branch placed **before** the CI gate so unrelated cancelled/failed sibling checks can't suppress the retry. For a marker at the **current head**: if already reviewed at head → skip (resolved); if `reset` elapsed (or unparseable → fail-open) → re-dispatch via the **normal trigger (never `force_review`)** so the advisory gate stays armed; if `reset` is still in the future → defer. Dispatch logic refactored into a shared `dispatch_review` helper (honours `DRY_RUN`/`MAX_DISPATCH`).

### Acceptance criteria
- **AC1** rate-limited run leaves a machine-detectable `status=rate-limited` marker — ✅ (`maybe_post_rate_limited_marker`, wired into the withhold path)
- **AC2** auto re-review after reset, no manual `force_review` — ✅ (sweep re-dispatches normal trigger; gate's head-age timeout then lets approval proceed once the window passes)
- **AC3** sibling cancelled/failed checks don't suppress the retry — ✅ (retry branch runs before the CI gate; tests cover FAILING + CANCELLED)
- **AC4** default behaviour for non-rate-limited PRs unchanged — ✅ (detect returns 1; existing stuck-green path intact; all prior tests still pass)
- **AC5** bats coverage — ✅ (see below)

### Tests (TDD — written red first, then green)
- `tests/test_sweep_stuck_reviews.bats`: +7 tests (elapsed-reset dispatch & no force_review; FAILING sibling AC3; CANCELLED sibling; future-reset deferral; reviewed-at-head resolution; stale-head no-op; DRY_RUN). 21/21 pass.
- `tests/dev-lead/unit/test_advisory_review_gate.bats`: +11 tests (`detect_advisory_rate_limit` true/false incl. latest-wins both directions, non-bot author, coderabbit, empty input; `maybe_post_rate_limited_marker` post + dedup). 50/50 pass.

### Verification
- `bats tests/test_sweep_stuck_reviews.bats` → 21/21
- `bats tests/dev-lead/unit/` → 485/485
- full lint.yml bats list (18 files) → 293/293
- `bash .dev-lead/scripts/dev-lead-lint.sh` → all checks passed (shellcheck on 51 files + 4 agent profiles)
- `shellcheck` on the three changed scripts → only pre-existing SC1091/SC2016 info findings; nothing new

### Notes / trade-offs
- A false positive (a gated bot's latest message coincidentally matches a rate-limit phrase while another bot is still withholding) only causes a **delayed** re-review — the re-dispatched run re-evaluates the gate and never approves prematurely.
- No new test files were added, so no `lint.yml` test-list change is required (both edited files are already covered).

--- a/scripts/lib/advisory-review-gate.sh
+++ b/scripts/lib/advisory-review-gate.sh
@@ -112,6 +112,105 @@ format_bot_status() {
   esac
 }
 
+# ────────────────────────────────────────────────────────────────────
+# Rate-limit detection + marker (issue #711)
+#
+# When an advisory/review bot withholds a real review because it is out of
+# quota, the gate keeps deferring approval but no event ever re-fires once CI
+# settles green — the PR strands at REVIEW_REQUIRED. detect_advisory_rate_limit
+# recognises that state and maybe_post_rate_limited_marker stamps a
+# machine-detectable marker so pr-review-sweep can auto-retry after the limit
+# resets (no manual force_review, which would bypass the gate entirely).
+# ────────────────────────────────────────────────────────────────────
+
+# Bots whose out-of-quota notice should arm the sweep retry. Superset of the
+# gate's ADVISORY_BOTS — adds coderabbitai, which posts an explicit rate-limit
+# comment but is not one of the bots the gate blocks on.
+# shellcheck disable=SC2034
+declare -ar RATE_LIMIT_NOTICE_BOTS=(
+  gemini-code-assist
+  copilot-pull-request-reviewer
+  sonarqubecloud
+  chatgpt-codex-connector
+  coderabbitai
+)
+
+# Case-insensitive phrases that indicate a bot is itself rate-limited / out of
+# quota (not merely discussing rate limiting). Kept reasonably specific so a
+# genuine review that mentions "rate limit" in passing does not arm a retry.
+_advisory_rate_limit_pattern() {
+  printf '%s' 'usage limit|rate.?limit|too many requests|quota (exceeded|reached|exhausted)|out of (quota|credits|tokens|requests)|limit (reached|exceeded|exhausted)|(reached|exceeded|hit) (the |your )?(usage |rate |daily |monthly )?limit'
+}
+
+# detect_advisory_rate_limit <reviews-comments-json>
+#   Returns 0 when a known advisory/review bot's LATEST submission body matches
+#   the rate-limit pattern; 1 otherwise. Only the latest submission per bot is
+#   considered, so a newer real review supersedes an older rate-limit notice
+#   (and vice versa). Non-bot authors are ignored.
+detect_advisory_rate_limit() {
+  local json="${1:-}"
+  [[ -z "$json" ]] && return 1
+
+  local bot_array pattern matched
+  bot_array=$(printf '%s\n' "${RATE_LIMIT_NOTICE_BOTS[@]}" | jq -R . | jq -s .)
+  pattern=$(_advisory_rate_limit_pattern)
+
+  matched=$(jq -r --argjson bots "$bot_array" --arg pat "$pattern" '
+    (
+      [(.reviews // [])[]  | {bot: .author.login, time: .submittedAt, body: (.body // "")}] +
+      [(.comments // [])[] | {bot: .author.login, time: .createdAt,   body: (.body // "")}]
+    )
+    | map(select([.bot] | inside($bots)))
+    | group_by(.bot)
+    | map(sort_by(.time) | last)
+    | map(select(.body | test($pat; "i")))
+    | length
+  ' <<< "$json" 2>/dev/null) || return 1
+
+  [[ "${matched:-0}" -gt 0 ]]
+}
+
+# maybe_post_rate_limited_marker <pr_url> <head_sha> <reset_iso> <comments_json>
+#   Posts a deduplicated rate-limited marker comment so the sweep can detect the
+#   withheld-due-to-rate-limit state and re-trigger a review after <reset_iso>.
+#   <comments_json> is the already-fetched PR snapshot (.comments[]) used for the
+#   dedup check, so no extra API call is needed to decide whether to post.
+#   The marker prefix ("rate-limited" before v1) deliberately never matches the
+#   idempotency marker regex (<!-- pr-review-agent v1 sha=...).
+maybe_post_rate_limited_marker() {
+  local pr_url="${1:-}" head_sha="${2:-}" reset_iso="${3:-}" comments_json="${4:-}"
+  if [[ -z "$pr_url" || -z "$head_sha" ]]; then
+    log_warn "maybe_post_rate_limited_marker: pr_url and head_sha are required"
+    return 1
+  fi
+
+  local marker="<!-- pr-review-agent rate-limited v1 sha=${head_sha} status=rate-limited reset=${reset_iso} -->"
+
+  # Dedup: skip if a rate-limited marker already exists at this exact head.
+  local cj="$comments_json"
+  [[ -z "$cj" ]] && cj='{}'
+  local already
+  already=$(jq -r --arg sha "$head_sha" '
+    [ (.comments // [])[]
+      | (.body // "")
+      | select(test("<!-- pr-review-agent rate-limited v1 sha=" + $sha + " ")) ]
+    | length' <<< "$cj" 2>/dev/null || echo 0)
+  if [[ "${already:-0}" -gt 0 ]]; then
+    log_info "Rate-limited marker already present at head ${head_sha:0:8} — not re-posting"
+    return 0
+  fi
+
+  local body="${marker}
+Advisory bots were rate-limited; auto-approval is withheld until they recover. pr-review-sweep will re-review this PR after ${reset_iso:-the limit resets}."
+
+  if gh pr comment "$pr_url" --body "$body" >/dev/null 2>&1; then
+    log_info "Posted rate-limited marker on $pr_url (head ${head_sha:0:8}, reset ${reset_iso:-n/a})"
+  else
+    log_warn "Failed to post rate-limited marker on $pr_url"
+    return 1
+  fi
+}
+
 # Instant (non-blocking) check of advisory bot status
 #
 # DESIGN: This uses a re-trigger pattern instead of blocking waits:

--- a/scripts/lib/advisory-review-gate.sh
+++ b/scripts/lib/advisory-review-gate.sh
@@ -160,7 +160,7 @@ detect_advisory_rate_limit() {
       [(.reviews // [])[]  | {bot: .author.login, time: .submittedAt, body: (.body // "")}] +
       [(.comments // [])[] | {bot: .author.login, time: .createdAt,   body: (.body // "")}]
     )
-    | map(select([.bot] | inside($bots)))
+    | map(select(.bot as $b | $bots | any(. == $b)))
     | group_by(.bot)
     | map(sort_by(.time) | last)
     | map(select(.body | test($pat; "i")))

--- a/scripts/review-one-pr.sh
+++ b/scripts/review-one-pr.sh
@@ -163,6 +163,19 @@ fi
       # even when advisory bots haven't submitted yet, so the force-review path works.
       echo "    force-review: advisory bots not all submitted, but FORCE_REVIEW=true — proceeding"
     else
+      # If the withhold is because an advisory/review bot is itself rate-limited
+      # (out of quota), stamp a machine-detectable marker so pr-review-sweep can
+      # auto-retry once the limit resets (issue #711). Without this, no bot event
+      # re-fires and the PR strands at REVIEW_REQUIRED with green CI. The marker
+      # is posted in an isolated subshell so the gate's helpers/vars do not leak.
+      (
+        # shellcheck source=lib/advisory-review-gate.sh
+        source "$SCRIPT_DIR/lib/advisory-review-gate.sh"
+        if detect_advisory_rate_limit "$PR_SNAPSHOT"; then
+          reset_iso=$(date -u -d "+${RATE_LIMIT_RETRY_BACKOFF:-3600} seconds" +%Y-%m-%dT%H:%M:%SZ 2>/dev/null || true)
+          maybe_post_rate_limited_marker "$PR_URL" "$PR_HEAD_SHA" "$reset_iso" "$PR_SNAPSHOT" || true
+        fi
+      ) || true
       # Bots are still reviewing — skip this run, will re-check on next bot review submission
       # Exit 100 (no-op) prevents workflow from consuming budget while awaiting bots
       echo "    skip: advisory bots still reviewing (non-blocking, will re-check on bot submission)"

--- a/scripts/review-one-pr.sh
+++ b/scripts/review-one-pr.sh
@@ -172,7 +172,9 @@ fi
         # shellcheck source=lib/advisory-review-gate.sh
         source "$SCRIPT_DIR/lib/advisory-review-gate.sh"
         if detect_advisory_rate_limit "$PR_SNAPSHOT"; then
-          reset_iso=$(date -u -d "+${RATE_LIMIT_RETRY_BACKOFF:-3600} seconds" +%Y-%m-%dT%H:%M:%SZ 2>/dev/null || true)
+          reset_iso=$(date -u -d "+${RATE_LIMIT_RETRY_BACKOFF:-3600} seconds" +%Y-%m-%dT%H:%M:%SZ 2>/dev/null \
+            || date -u -v+"${RATE_LIMIT_RETRY_BACKOFF:-3600}"S +%Y-%m-%dT%H:%M:%SZ 2>/dev/null \
+            || true)
           maybe_post_rate_limited_marker "$PR_URL" "$PR_HEAD_SHA" "$reset_iso" "$PR_SNAPSHOT" || true
         fi
       ) || true

--- a/scripts/sweep-stuck-reviews.sh
+++ b/scripts/sweep-stuck-reviews.sh
@@ -156,7 +156,7 @@ while IFS= read -r pr_url; do
   rl_reset=$(jq -r --arg sha "$head_sha" '
     [ ((.reviews // []) + (.comments // []))[]
       | (.body // "")
-      | select(test("<!-- pr-review-agent rate-limited v1 sha=" + $sha + " "))
+      | select(test("<!-- pr-review-agent rate-limited v1 sha=" + $sha + " ") and test("reset=[^ ]+"))
       | capture("reset=(?<r>[^ ]+)") | .r ]
     | last // ""' <<< "$snapshot" 2>/dev/null || echo "")
   if [ -n "$rl_reset" ]; then
@@ -165,7 +165,9 @@ while IFS= read -r pr_url; do
       echo "  skip $pr_url — rate-limited marker present but already reviewed at head ${head_sha:0:8}"
       continue
     fi
-    reset_epoch=$(date -u -d "$rl_reset" +%s 2>/dev/null || echo "")
+    reset_epoch=$(date -u -d "$rl_reset" +%s 2>/dev/null \
+      || date -u -j -f "%Y-%m-%dT%H:%M:%SZ" "$rl_reset" +%s 2>/dev/null \
+      || echo "")
     now_epoch=$(date -u +%s)
     # Dispatch once the reset has elapsed; an unparseable reset fails open to a
     # retry so a malformed marker can never strand the PR indefinitely.

--- a/scripts/sweep-stuck-reviews.sh
+++ b/scripts/sweep-stuck-reviews.sh
@@ -92,6 +92,26 @@ inspected=0
 stuck=0
 dispatched=0
 
+# dispatch_review <pr_url> <label>
+# Re-dispatch a review through the normal trigger (never force_review — the
+# advisory gate must stay armed). Honours DRY_RUN and updates the counters.
+dispatch_review() {
+  local _url="$1" _label="$2"
+  stuck=$((stuck + 1))
+  echo "  $_label: $_url — needs re-review"
+  if [ "$DRY_RUN_BOOL" = "true" ]; then
+    echo "    dry-run: would dispatch $TRIGGER_WORKFLOW for $_url"
+    return 0
+  fi
+  if gh workflow run "$TRIGGER_WORKFLOW" --repo "$AGENT_REPO" "${REF_FLAGS[@]}" \
+       -f pr_url="$_url"; then
+    dispatched=$((dispatched + 1))
+    echo "    dispatched review for $_url"
+  else
+    echo "::warning::sweep: failed to dispatch review for $_url"
+  fi
+}
+
 while IFS= read -r pr_url; do
   [ -z "$pr_url" ] && continue
   if [ "$dispatched" -ge "$MAX_DISPATCH" ]; then
@@ -112,12 +132,6 @@ while IFS= read -r pr_url; do
     continue
   fi
 
-  ci_status=$(compute_ci_status "$(jq '.statusCheckRollup' <<< "$snapshot")")
-  if [ "$ci_status" != "passing" ]; then
-    echo "  skip $pr_url — CI '$ci_status' (not green yet)"
-    continue
-  fi
-
   head_sha=$(jq -r '.headRefOid // ""' <<< "$snapshot")
   if [ -z "$head_sha" ]; then
     echo "  skip $pr_url — head SHA is empty"
@@ -132,26 +146,49 @@ while IFS= read -r pr_url; do
       | (.body // "")
       | select(test("<!-- pr-review-agent v1 sha=" + $sha + " ")) ]
     | length' <<< "$snapshot" 2>/dev/null || echo 0)
+
+  # Rate-limited withhold retry (issue #711). A pr-review run that withheld
+  # approval because advisory bots were rate-limited stamps a distinct marker:
+  #   <!-- pr-review-agent rate-limited v1 sha=<HEAD> status=rate-limited reset=<ISO> -->
+  # Handle it BEFORE the CI gate so unrelated cancelled/failed sibling checks
+  # cannot suppress the retry (AC3). Extract the reset for a marker at the
+  # current head (sha followed by a space, never a prefix match).
+  rl_reset=$(jq -r --arg sha "$head_sha" '
+    [ ((.reviews // []) + (.comments // []))[]
+      | (.body // "")
+      | select(test("<!-- pr-review-agent rate-limited v1 sha=" + $sha + " "))
+      | capture("reset=(?<r>[^ ]+)") | .r ]
+    | last // ""' <<< "$snapshot" 2>/dev/null || echo "")
+  if [ -n "$rl_reset" ]; then
+    if [ "${reviewed_at_head:-0}" -gt 0 ]; then
+      # A real review already landed at this head — rate-limit state is resolved.
+      echo "  skip $pr_url — rate-limited marker present but already reviewed at head ${head_sha:0:8}"
+      continue
+    fi
+    reset_epoch=$(date -u -d "$rl_reset" +%s 2>/dev/null || echo "")
+    now_epoch=$(date -u +%s)
+    # Dispatch once the reset has elapsed; an unparseable reset fails open to a
+    # retry so a malformed marker can never strand the PR indefinitely.
+    if [ -z "$reset_epoch" ] || [ "$now_epoch" -ge "$reset_epoch" ]; then
+      dispatch_review "$pr_url" "rate-limited (reset $rl_reset elapsed, head ${head_sha:0:8})"
+    else
+      echo "  defer $pr_url — rate-limited until $rl_reset (head ${head_sha:0:8})"
+    fi
+    continue
+  fi
+
+  ci_status=$(compute_ci_status "$(jq '.statusCheckRollup' <<< "$snapshot")")
+  if [ "$ci_status" != "passing" ]; then
+    echo "  skip $pr_url — CI '$ci_status' (not green yet)"
+    continue
+  fi
+
   if [ "${reviewed_at_head:-0}" -gt 0 ]; then
     echo "  skip $pr_url — already reviewed at head ${head_sha:0:8}"
     continue
   fi
 
-  stuck=$((stuck + 1))
-  echo "  stuck-green: $pr_url (head ${head_sha:0:8}) — needs re-review"
-
-  if [ "$DRY_RUN_BOOL" = "true" ]; then
-    echo "    dry-run: would dispatch $TRIGGER_WORKFLOW for $pr_url"
-    continue
-  fi
-
-  if gh workflow run "$TRIGGER_WORKFLOW" --repo "$AGENT_REPO" "${REF_FLAGS[@]}" \
-       -f pr_url="$pr_url"; then
-    dispatched=$((dispatched + 1))
-    echo "    dispatched review for $pr_url"
-  else
-    echo "::warning::sweep: failed to dispatch review for $pr_url"
-  fi
+  dispatch_review "$pr_url" "stuck-green (head ${head_sha:0:8})"
 done < "$candidates_file"
 
 echo ""

--- a/tests/dev-lead/unit/test_advisory_review_gate.bats
+++ b/tests/dev-lead/unit/test_advisory_review_gate.bats
@@ -443,3 +443,132 @@ MOCK_EOF
   "
   [[ "$output" == *"CHANGES_REQUESTED"* ]]
 }
+
+# ────────────────────────────────────────────────────────────────────
+# RATE-LIMIT DETECTION TESTS (issue #711)
+#
+# detect_advisory_rate_limit <reviews-comments-json>
+#   returns 0 when a known advisory/review bot's LATEST submission body matches
+#   a rate-limit phrase; 1 otherwise. The marker it arms lets pr-review-sweep
+#   auto-retry once the limit resets (no manual force_review).
+# ────────────────────────────────────────────────────────────────────
+
+@test "detect_advisory_rate_limit: function exists" {
+  grep -q "detect_advisory_rate_limit()" "$SCRIPT_DIR/lib/advisory-review-gate.sh"
+}
+
+@test "detect_advisory_rate_limit: true when codex posts a usage-limit notice" {
+  local json
+  json='{"reviews":[],"comments":[{"author":{"login":"chatgpt-codex-connector"},"createdAt":"2026-06-07T10:00:00Z","body":"You have reached your usage limit. Please try again later."}]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 0 ]
+}
+
+@test "detect_advisory_rate_limit: true when coderabbit posts a rate-limit notice" {
+  local json
+  json='{"reviews":[{"author":{"login":"coderabbitai"},"state":"COMMENTED","submittedAt":"2026-06-07T10:00:00Z","body":"Review skipped: rate limit exceeded. Reviews will resume after the limit resets."}],"comments":[]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 0 ]
+}
+
+@test "detect_advisory_rate_limit: false when bot's latest submission is a real review" {
+  local json
+  json='{"reviews":[{"author":{"login":"chatgpt-codex-connector"},"state":"COMMENTED","submittedAt":"2026-06-07T10:00:00Z","body":"I found a potential null pointer dereference on line 42."}],"comments":[]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 1 ]
+}
+
+@test "detect_advisory_rate_limit: false when a newer real review supersedes the rate-limit notice" {
+  # Older comment is a rate-limit notice; newer review is a real review.
+  # The latest submission per bot wins → not rate-limited.
+  local json
+  json='{"reviews":[{"author":{"login":"chatgpt-codex-connector"},"state":"COMMENTED","submittedAt":"2026-06-07T11:00:00Z","body":"LGTM, no issues found."}],"comments":[{"author":{"login":"chatgpt-codex-connector"},"createdAt":"2026-06-07T10:00:00Z","body":"You have reached your usage limit."}]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 1 ]
+}
+
+@test "detect_advisory_rate_limit: true when a newer rate-limit notice supersedes an older real review" {
+  # Older review was real; newer comment is a rate-limit notice → rate-limited now.
+  local json
+  json='{"reviews":[{"author":{"login":"chatgpt-codex-connector"},"state":"COMMENTED","submittedAt":"2026-06-07T10:00:00Z","body":"Looks good overall."}],"comments":[{"author":{"login":"chatgpt-codex-connector"},"createdAt":"2026-06-07T11:00:00Z","body":"You have reached your usage limit. Try again later."}]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 0 ]
+}
+
+@test "detect_advisory_rate_limit: false when the rate-limit phrase comes from a non-advisory (human) author" {
+  local json
+  json='{"reviews":[],"comments":[{"author":{"login":"some-human"},"createdAt":"2026-06-07T10:00:00Z","body":"We should add a rate limit to this endpoint, quota exceeded errors are common."}]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '$json'"
+  [ "$status" -eq 1 ]
+}
+
+@test "detect_advisory_rate_limit: false on empty input" {
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run bash -c "source '$gate_script'; detect_advisory_rate_limit '{\"reviews\":[],\"comments\":[]}'"
+  [ "$status" -eq 1 ]
+}
+
+# ────────────────────────────────────────────────────────────────────
+# RATE-LIMITED MARKER POSTING TESTS (issue #711)
+#
+# maybe_post_rate_limited_marker <pr_url> <head_sha> <reset_iso> <comments_json>
+#   posts a deduplicated marker comment so the sweep can detect the
+#   withheld-due-to-rate-limit state and auto-retry after <reset_iso>.
+# ────────────────────────────────────────────────────────────────────
+
+@test "maybe_post_rate_limited_marker: function exists" {
+  grep -q "maybe_post_rate_limited_marker()" "$SCRIPT_DIR/lib/advisory-review-gate.sh"
+}
+
+@test "maybe_post_rate_limited_marker: posts a marker comment when none exists at head" {
+  local tmpdir
+  tmpdir=$(mktemp -d)
+  cat > "$tmpdir/gh" << MOCK_EOF
+#!/usr/bin/env bash
+printf '%s\n' "\$*" >> "$tmpdir/calls"
+exit 0
+MOCK_EOF
+  chmod +x "$tmpdir/gh"
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run env PATH="$tmpdir:$PATH" bash -c "
+    source '$gate_script'
+    maybe_post_rate_limited_marker 'https://github.com/owner/repo/pull/123' 'abc123' '2999-01-01T00:00:00Z' '{\"comments\":[]}'
+  "
+  local calls; calls=$(cat "$tmpdir/calls" 2>/dev/null || true)
+  rm -rf "$tmpdir"
+  [ "$status" -eq 0 ]
+  # gh pr comment was invoked with the rate-limited marker for this head
+  [[ "$calls" == *"pr comment"* ]]
+  [[ "$calls" == *"rate-limited v1 sha=abc123"* ]]
+  [[ "$calls" == *"status=rate-limited"* ]]
+  [[ "$calls" == *"reset=2999-01-01T00:00:00Z"* ]]
+}
+
+@test "maybe_post_rate_limited_marker: does not re-post when a marker already exists at head" {
+  local tmpdir
+  tmpdir=$(mktemp -d)
+  cat > "$tmpdir/gh" << MOCK_EOF
+#!/usr/bin/env bash
+printf '%s\n' "\$*" >> "$tmpdir/calls"
+exit 0
+MOCK_EOF
+  chmod +x "$tmpdir/gh"
+  local existing
+  existing='{"comments":[{"body":"<!-- pr-review-agent rate-limited v1 sha=abc123 status=rate-limited reset=2999-01-01T00:00:00Z -->\n\nAdvisory bots were rate-limited."}]}'
+  local gate_script="$SCRIPT_DIR/lib/advisory-review-gate.sh"
+  run env PATH="$tmpdir:$PATH" bash -c "
+    source '$gate_script'
+    maybe_post_rate_limited_marker 'https://github.com/owner/repo/pull/123' 'abc123' '2999-01-01T00:00:00Z' '$existing'
+  "
+  local calls; calls=$(cat "$tmpdir/calls" 2>/dev/null || true)
+  rm -rf "$tmpdir"
+  [ "$status" -eq 0 ]
+  # No new comment should be posted when a marker for this head already exists
+  [[ "$calls" != *"pr comment"* ]]
+}

--- a/tests/test_sweep_stuck_reviews.bats
+++ b/tests/test_sweep_stuck_reviews.bats
@@ -208,3 +208,96 @@ url_for() { echo "https://github.com/petry-projects/demo/pull/$1"; }
   [ "$status" -eq 0 ]
   ! grep -qF -- "--ref" "$GH_LOG"
 }
+
+# ───────────────────────────────────────────────────────────────────────────
+# Rate-limited withhold retry (issue #711)
+#
+# A pr-review run that withholds approval because advisory bots were
+# rate-limited leaves a distinct marker:
+#   <!-- pr-review-agent rate-limited v1 sha=<HEAD> status=rate-limited reset=<ISO> -->
+# Once the embedded reset has passed (and the PR has NOT since been reviewed at
+# head), the sweep re-dispatches it — even when the CI rollup is not "passing"
+# (unrelated cancelled/failed sibling checks must not suppress the retry).
+# ───────────────────────────────────────────────────────────────────────────
+
+# rl_comment <head> <reset-iso>: a rate-limited withhold marker comment body.
+rl_comment() {
+  printf '[{"body":"<!-- pr-review-agent rate-limited v1 sha=%s status=rate-limited reset=%s -->\\n\\nAdvisory bots were rate-limited."}]' "$1" "$2"
+}
+PAST_RESET='2000-01-01T00:00:00Z'
+FUTURE_RESET='2999-01-01T00:00:00Z'
+
+@test "rate-limited marker with elapsed reset is re-dispatched (CI passing)" {
+  write_pr 711 "REVIEW_REQUIRED" "$ROLLUP_PASS" "rl711" "[]" "$(rl_comment rl711 "$PAST_RESET")"
+  url_for 711 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  grep -qF -- "-f pr_url=$(url_for 711)" "$GH_LOG"
+  # Retry uses the normal trigger path — never force_review (the advisory gate
+  # must stay armed so we never approve while bots are still rate-limited).
+  ! grep -qF -- "force_review" "$GH_LOG"
+}
+
+@test "rate-limited retry fires even when a sibling check is FAILING (issue #711 AC3)" {
+  write_pr 712 "REVIEW_REQUIRED" "$ROLLUP_FAIL" "rl712" "[]" "$(rl_comment rl712 "$PAST_RESET")"
+  url_for 712 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  grep -qF -- "-f pr_url=$(url_for 712)" "$GH_LOG"
+}
+
+@test "rate-limited retry fires even when a sibling check is CANCELLED" {
+  local rollup_cancelled='[{"name":"CI","status":"COMPLETED","conclusion":"SUCCESS"},{"name":"dev-lead / dispatch","status":"COMPLETED","conclusion":"CANCELLED"}]'
+  write_pr 713 "REVIEW_REQUIRED" "$rollup_cancelled" "rl713" "[]" "$(rl_comment rl713 "$PAST_RESET")"
+  url_for 713 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  grep -qF -- "-f pr_url=$(url_for 713)" "$GH_LOG"
+}
+
+@test "rate-limited marker whose reset is still in the future is NOT re-dispatched" {
+  write_pr 714 "REVIEW_REQUIRED" "$ROLLUP_PASS" "rl714" "[]" "$(rl_comment rl714 "$FUTURE_RESET")"
+  url_for 714 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_LOG" ]
+}
+
+@test "rate-limited marker is ignored once the PR has been reviewed at head" {
+  # An idempotency marker at the same head means a real review already landed —
+  # the rate-limited state is resolved, so the retry must NOT fire.
+  local comments
+  comments='[{"body":"<!-- pr-review-agent rate-limited v1 sha=rl715 status=rate-limited reset=2000-01-01T00:00:00Z -->"},{"body":"<!-- pr-review-agent v1 sha=rl715 --> reviewed"}]'
+  write_pr 715 "REVIEW_REQUIRED" "$ROLLUP_PASS" "rl715" "[]" "$comments"
+  url_for 715 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_LOG" ]
+}
+
+@test "rate-limited marker for a DIFFERENT (stale) head does not trigger a retry" {
+  # Marker is for an old head; current head has no marker and CI is failing, so
+  # neither the rate-limited branch nor the normal stuck-green branch fires.
+  write_pr 716 "REVIEW_REQUIRED" "$ROLLUP_FAIL" "rl716NEW" "[]" "$(rl_comment rl716OLD "$PAST_RESET")"
+  url_for 716 > "$SWEEP_PRS_FILE"
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_LOG" ]
+}
+
+@test "rate-limited retry honours DRY_RUN" {
+  write_pr 717 "REVIEW_REQUIRED" "$ROLLUP_FAIL" "rl717" "[]" "$(rl_comment rl717 "$PAST_RESET")"
+  url_for 717 > "$SWEEP_PRS_FILE"
+  export DRY_RUN=true
+
+  run bash "$SCRIPT"
+  [ "$status" -eq 0 ]
+  [ ! -s "$GH_LOG" ]
+  [[ "$output" == *"$(url_for 717)"* ]]
+}

